[AIRFLOW-3504] Extend/refine the functionality of "/health" endpoint

XD-DENG opened the pull request:
### Jira

  - https://issues.apache.org/jira/browse/AIRFLOW-3504


### Description

**=== UPDATED BASED ON THE LATEST COMMIT ===**

#### The Issue That This PR Tried to Address

For both www/ and www_rbac/, there is "/health" endpoint, which is supposed to help users know the "health" condition of the Airflow application.

But the functionality of it at this moment is very limited. It can only help check if the webserver is running, but actually the "health" of the application may be bad even though this endpoint returns "The server is healthy!" (scheduler may be down, database backend may be disconnected, etc)

#### What Have Been Done in This PR

Extend the functionality of "`/health`" endpoint:

1. Check if database backend can be connected;
2. Check the latest scheduler heartbeat.

This is done for both `/www` and `/www_rbac`. Tests are added accordingly as well.

Note that no authentication is required to access "`/health`" endpoint (no sensitive information will be exposed through it).

(Deleted an unnecessary line in `airflow/www/views.py` as well)

#### How Final Result Looks Like

The "`/health`" endpoint will return a JSON like

```
  {
    "metadatabase":{
      "status":"healthy"
    },
    "scheduler":{
      "status":"healthy",
      "latest_scheduler_heartbeat":"2018-12-26 17:15:11+00:00"
    }
  }
```

(Currently it simply returns "`The server is healthy!`", which is not very useful, and can't really check the health condition of the application)

#### Others

**No authentication is required to access this `/health` endpoint. This is a conscious decision.**

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Changed tests below accordingly:
- `tests/core.py` (for /www)
- `tests/www_rbac/test_views.py` (for /www_rbac)

### Code Quality

- [x] Passes `flake8`

ashb commented on airflow/www/blueprints.py:
```diff
@@ -32,6 +35,22 @@ def index():
+    try:
+        session.query(DM).count()
+    except Exception as _:
+        payload['database'] = "unavailable"
```
Should this return something other than a 200 in this case? I'm not suer what is "typical" for a health check URL.

XD-DENG replied:
Thanks @ashb .

I also noticed that people use status code to indicate the health check result in some applications. But the case here may be a bit different: Let's say webserver & database connection are both healthy, while scheduler is not working at all. What status code should we give?

So I would suggest to just give 200, and simply use the contents of the response to indicate the health of each component. Do you think it's ok?

ashb replied:
I can see arguments both ways, but which ever we choose needs to be clearly documented, and also probably make it clear this is not designed to be hooked up to a Kubernetes healthcheck (as it will never return anything other than 200< so won't ever "fail" as far as kube is concerned)

feng-tao commented on airflow/www_rbac/views.py:
```diff
@@ -146,6 +146,29 @@ def render(self, template, **context):
 
 
 class Airflow(AirflowBaseView):
+    @expose('/health')
```
do we need access control for this endpoint? I assume only admin should access this endpoint?

XD-DENG replied:
Thanks @feng-tao  for the review!

The current implementation is not requiring any logging-in to access `/health` endpoint. But of course it only returns a very "safe" response "`The server is healthy!`" (which is not useful/accurate).

For this PR, I would still like to suggest not require admin or logging-in to access this endpoint. This is just like the status page of GitHub. Everyone should be able access it without any logging-in. But on the other hand, we do need to be careful of what to expose via this endpoint. This was partially covered in my last reply comment to Kaxil.

feng-tao commented on airflow/www/views.py:
```diff
@@ -377,7 +377,6 @@ def index(self):
     @expose('/chart_data')
     @data_profiling_required
     @wwwutils.gzipped
-    # @cache.cached(timeout=3600, key_prefix=wwwutils.make_cache_key)
```
why this line remove from this pr?

XD-DENG replied:
This line was commented, and no import statement for `cache`.

I believe the original author tried to use `Flask-Cache` to improve performance, but eventually related codes were removed, while this line was missed in the removing process.

XD-DENG replied:
I deleted it here "by the way", since it may not be worthwhile to have a separate PR for it.

feng-tao reviewed (commented):
I think the high level concept lgtm. I will look at the pr in more detail for the implementation once back to office next week.

feng-tao commented on airflow/www/blueprints.py:
```diff
@@ -32,6 +37,26 @@ def index():
 
 @routes.route('/health')
 def health():
```
it would be good to add a comment.

XD-DENG replied:
Addressed.

feng-tao commented on airflow/www/blueprints.py:
```diff
@@ -32,6 +37,26 @@ def index():
+    if not latest_scheduler_run:
+        scheduler_status = 'unhealthy'
+    else:
+        if timezone.utcnow() - latest_scheduler_run <= timedelta(seconds=30):
```
could you comment on this 30 seconds magic number and why we pick 30 seconds to define the health/non-health for scheduler?

XD-DENG replied:
I have put this threshold into `airflow.cfg` as a configuration item, with default value 30 seconds. 

Users can change this value based on their needs.

feng-tao commented on airflow/www/blueprints.py:
```diff
@@ -32,6 +37,26 @@ def index():
+            scheduler_status = 'unhealthy'
+
+    payload['scheduler'] = {'status': scheduler_status,
+                            'latest_scheduler_run': str(latest_scheduler_run)}
```
what is the reason behind converting the datetime object to a string type?

XD-DENG replied:
Thanks @feng-tao for reviewing.

Yes, this converting was added purposely.

```python
import json

from airflow.www import utils as wwwutils
from airflow.utils import timezone


ts = timezone.utcnow()

print(json.loads(wwwutils.json_response({"timestamp":ts}).data))
print(json.loads(wwwutils.json_response({"timestamp":str(ts)}).data))
```

**Result:**
```
{'timestamp': '2019-01-06T09:39:49Z'}
{'timestamp': '2019-01-06 09:39:49.894382+00:00'}
```

The second format is more favourable to me. That was why I purposely chose explicit str conversion.

For other comments, please allow me a bit more time to check & address. Thanks again!

feng-tao commented on airflow/www_rbac/blueprints.py:
```diff
@@ -17,19 +17,11 @@
-
-
-@routes.route('/health')
-def health():
```
I wonder why we modify the health function in the blueprint.py in the old UI while delete this function and move it to views.py?

Either way is fine, just want to understand why we do the same(both delete and move to views.py) for old/new UI?

XD-DENG replied:
I have tried: for the old UI, if we put this function into `views.py`, the endpoint will be something like `https://<host>:8080/admin/airflow/health`, while I believe it's better to make it `https://<host>:8080/health` to be consistent with the new UI.

feng-tao commented on airflow/www_rbac/views.py:
```diff
@@ -147,6 +147,33 @@ def render(self, template, **context):
+    @expose('/health')
+    @provide_session
+    def health(self, session=None):
+        from sqlalchemy import func
```
why do local import instead of global import for the function? Do we have namespace conflict?

XD-DENG replied:
Addressed. Now imported at the top.

feng-tao commented on airflow/www_rbac/views.py:
```diff
@@ -147,6 +147,33 @@ def render(self, template, **context):
+        if not latest_scheduler_run:
+            scheduler_status = 'unhealthy'
+        else:
+            if timezone.utcnow() - latest_scheduler_run <= timedelta(seconds=30):
```
same comment on why we pick 30 seconds as magic number(needs some comments). And it would be good if we put it in a config value so that it allows users to change.

XD-DENG replied:
Addressed. Now it's a configuration item

feng-tao reviewed (commented):
left some comments. thanks @XD-DENG

XD-DENG committed:
[AIRFLOW-3504] Refine the functionality of "/health" endpoint

Extend the functionality of "/health" endpoint by:
1. Checking if database backend can be connected;
2. Checking the latest scheduler heartbeat

The response will be in format:
  {
    "metadatabase":{
      "status":"healthy"
    },
    "scheduler":{
      "status":"healthy",
      "latest_scheduler_heartbeat":"2018-12-26 17:15:11+00:00"
    }
  }

This is done for both /www and /www_rbac.

No authentication is required to access this endpoint (no
sensitive information will be exposed through it).

Tests & documentation are added accordingly.

(Deleted an unnecessary line in airflow/www/views.py as well)

feng-tao reviewed (approved):
LGTM. Thanks @XD-DENG .
